pxd_fastpath.c

```diff
@@ -526,6 +528,54 @@ void pxd_fastpath_adjust_limits(struct pxd_device *pxd_dev, struct request_queue
 	disableFastPath(pxd_dev, false);
 }
 
+// reset device called during device cleanup actions from any internal state.
+// consider node wipe, device remove while suspended etc.
+void pxd_fastpath_reset_device(struct pxd_device *pxd_dev)
+{
+	struct pxd_fastpath_extension *fp = &pxd_dev->fp;
+	struct pxd_context *ctx = pxd_dev->ctx;
+	struct fuse_conn *fc = &ctx->fc;
+	struct fuse_req *req;
+
+	if (!fastpath_enabled(pxd_dev)) {
+		return;
+	}
+
+	disableFastPath(pxd_dev, true);
+
+	// abort any inflight ioswitch
+	if (atomic_read(&fp->ioswitch_active)) {
+		req = request_find(fc, pxd_dev->fp.switch_uid);
+		if (!IS_ERR_OR_NULL(req)) {
+			// overwrite switch request to fail all pending IOs
+			req->in.h.opcode = PXD_FAILOVER_TO_USERSPACE;
+			req->out.h.error = -EIO; // force failure status
+			request_end(fc, req, true /* should lock */);
+		} else {
+			pxd_dev->fp.switch_uid = 0;
+			atomic_set(&fp->ioswitch_active, 0);
+		}
+	}
+
+	// resume from userspace IO suspends
+	pxd_request_resume(pxd_dev);
+
+	// resume all suspended callstacks
+	while (atomic_read(&fp->suspend) > 0) {
+		pxd_resume_io(pxd_dev);
+	}
+
+#ifdef __PXD_BIO_BLKMQ__
+	BUG_ON(atomic_read(&fp->blkmq_frozen) != 0);
+#endif
+	BUG_ON(atomic_read(&fp->suspend) != 0);
+	BUG_ON(atomic_read(&fp->app_suspend) != 0);
+	BUG_ON(atomic_read(&fp->ioswitch_active) != 0);
+	BUG_ON(pxd_dev->fp.switch_uid != 0);
```

Comment:
Can we please remove these BUG_ONs? We can keep them in master. At customer it's better to have a stuck device than a kernel panic (which may or may not save a bt).

Reply:
done